common/src/Model/Brush.cpp

```diff
@@ -466,12 +467,85 @@ namespace TrenchBroom {
             doSetNewGeometry(worldBounds, matcher, newGeometry);
         }
 
+        static vm::vec3 snapPoint(const vm::vec3& point, const FloatType snapToF) {
+            return snapToF * vm::round(point / snapToF);
+        }
+
+        /**
+         * Snaps all vertices.
+         */
         static BrushGeometry snappedGeometry(const BrushGeometry& geometry, const FloatType snapToF) {
             std::vector<vm::vec3> points;
             points.reserve(geometry.vertexCount());
             
             for (const auto* vertex : geometry.vertices()) {
-                points.push_back(snapToF * vm::round(vertex->position() / snapToF));
+                points.push_back(snapPoint(vertex->position(), snapToF));
+            }
+
+            return BrushGeometry(std::move(points));
+        }
+
+        /**
+         * Snaps only the given vertices.
+         */
+        static BrushGeometry snappedGeometry(const BrushGeometry& geometry, const FloatType snapToF, const std::vector<vm::vec3>& vertexPositionsToSnap) {
+            std::vector<vm::vec3> points;
+            points.reserve(geometry.vertexCount());
+
+            for (const auto* vertex : geometry.vertices()) {
+                if (kdl::vec_contains(vertexPositionsToSnap, vertex->position())) {
+                    points.push_back(snapPoint(vertex->position(), snapToF));
+                } else {
+                    points.push_back(vertex->position());
+                }
+            }
+
+            return BrushGeometry(std::move(points));
+        }
+
+        static bool shouldSnap(const std::vector<vm::segment3>& edgePositionsToSnap, const vm::vec3& vertexPosition) {
+            return std::any_of(edgePositionsToSnap.cbegin(), edgePositionsToSnap.cend(), [&](const vm::segment3& edge){
+                return edge.start() == vertexPosition || edge.end() == vertexPosition;
+            });
+        }
+
+        /**
+         * Snaps only the given edges.
+         */
+        static BrushGeometry snappedGeometry(const BrushGeometry& geometry, const FloatType snapToF, const std::vector<vm::segment3>& edgePositionsToSnap) {
```

Comment:
Why not implement these in terms of `static BrushGeometry snappedGeometry(const BrushGeometry& geometry, const FloatType snapToF, const std::vector<vm::vec3>& vertexPositionsToSnap)`? Just collect the vertices of the edges / faces to snap in a `kdl::vector_set`, and pass them on.